ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java

```diff
@@ -407,40 +427,93 @@ public void close() {
-          true);
+          false);
       diskIOScheduler.getSlice(BlobStoreStats.IO_SCHEDULER_JOB_TYPE, BlobStoreStats.IO_SCHEDULER_JOB_ID,
           indexEntries.size());
     } catch (IOException e) {
       throw new StoreException(
           String.format("I/O exception while getting entries from index segment for store %s", storeId), e,
           StoreErrorCodes.IOError);
     }
+    addPutEntriesForDelete(indexSegment.getStartOffset().getOffset(), indexEntries);
+    updateExpiryTimeForAllPuts(indexEntries);
     return indexEntries;
   }
 
+  /**
+   * Adds put entries for any delete entries that may have squashed put entries in the same segment. Requires that a
+   * put entry (if it exists) should occur before a delete entry for the same key in {@code indexEntries}
+   * @param indexSegmentStartOffset the start offset of the {@link IndexSegment} from which {@code indexEntries} was
+   *                                obtained
+   * @param indexEntries the list of {@link IndexEntry} from the same {@link IndexSegment}. This list will be sorted
+   *                     in the ascending order of key,offset
+   * @throws StoreException if there are problems accessing the index
+   */
+  private void addPutEntriesForDelete(long indexSegmentStartOffset, List<IndexEntry> indexEntries)
+      throws StoreException {
+    Set<StoreKey> seenPuts = new HashSet<>();
+    List<IndexEntry> newEntries = new ArrayList<>();
+    for (IndexEntry entry : indexEntries) {
+      IndexValue value = entry.getValue();
+      if (value.isFlagSet(IndexValue.Flags.Delete_Index)) {
+        if (value.getOriginalMessageOffset() != IndexValue.UNKNOWN_ORIGINAL_MESSAGE_OFFSET
+            && value.getOriginalMessageOffset() != value.getOffset().getOffset()
+            && value.getOriginalMessageOffset() >= indexSegmentStartOffset && !seenPuts.contains(entry.getKey())) {
+          // We need to find the original put (if it has not already been processed)
+          IndexValue originalPutValue = getPutRecordForDeletedKey(entry.getKey(), value);
+          newEntries.add(new IndexEntry(entry.getKey(), originalPutValue));
+        }
+      } else if (!entry.getValue().isFlagSet(IndexValue.Flags.Ttl_Update_Index)) {
+        seenPuts.add(entry.getKey());
+      }
+    }
+    indexEntries.addAll(newEntries);
+    indexEntries.sort(KEY_OFFSET_COMPARATOR);
+  }
+
+  /**
+   * Updates the expiry time (ms) for all PUT index values
+   * @param indexEntries the {@link IndexEntry} list to update the expiry time (ms) for
+   * @throws StoreException if there are problems interacting with the index
+   */
+  private void updateExpiryTimeForAllPuts(List<IndexEntry> indexEntries) throws StoreException {
+    for (IndexEntry entry : indexEntries) {
+      if (!entry.getValue().isFlagSet(IndexValue.Flags.Ttl_Update_Index) && !entry.getValue()
+          .isFlagSet(IndexValue.Flags.Delete_Index)) {
+        long expiryTimeMs = index.findKey(entry.getKey()).getExpiresAtMs();
```

Comment:
By default, most current use cases don't care about `TTL_UPDATE` values so it makes sense to have only those 2. Use cases that do need the `TTL_UPDATE` ask for it.

In fact, I could even make the argument that this function should always return the PUT value (unless otherwise specified) but with the delete flag updated if a delete has occurred.